### DISCUSSION TITLE
Momo "ART parameters are abnormal" from Lsposed 1.9.2

### DISCUSSION THREAD

**Discussion by bengalih** [LSPosed/LSPosed#2841] [Q&A]
-  Pixel 3 XL running stock A11 ROM.
- TJW Magisk 26.4 Zygisk enabled
- LSposed 1.9.2
- Momo - 4.3.1 and 4.4.1

In short, Momo 4.4.1 will report the following detail warning as soon as Lsposed is installed and the device rebooted:

"ART parameters are abnormal" 

Removing Lsposed does not stop this warning from showing.
In fact, even removing root (by reflashing the boot.img file) does not rid the device of this warning.

I can find nothing short of a full factory reset that removes it.

Can someone explain what LSposed is doing that triggers this modification and why even removing it does not revert.

This is referenced here:
https://github.com/orgs/LSPosed/discussions/2090

But claimed fixed in 1.9.0, which it clearly does not appear to be (or there was a regression).

**Comment by Howard20181** [comment]
For Android 10 and above, always keep SELinux enforced should fix it
For Android 9 and below there are no fixes that can be applied

**Reply by bengalih** [reply]
Please provide more information.  The result of "getenforce" on my device is "Enforcing".
So it would appear that my SELinux is enforcing and I am on Android 11.

I don't believe this should be closed as my issue is not resolved.

Please let me know what else I can provide.

**Reply by Howard20181** [reply]
Devices can't be fake enforced, can't trust the results of getenforce

This issue will not be fixed, it does not occur on normal standard AOSP devices

**Reply by bengalih** [reply]
How can you determine that my device is "fake enforced"?  
Also, as I stated Momo does NOT display this error without LSposed installed.
Even if LSposed is removed, the error remains.

So it is clear that the install of LSposed changes something that causes Momo to fail.
Why does the removal of LSposed not revert the changes?

thank you.

**Reply by Howard20181** [reply]
> How can you determine that my device is "fake enforced"?

Determined by the actual source code of the implementation

**Reply by bengalih** [reply]
I'm sorry, but why do you refuse to address the second part of the question?
The ART error is not present prior to LSposed being installed.
This means the installation of LSposed causes this error to be shown.
The removal of LSposed does not remove the error.
How is this explained?

thank you.

**Reply by Howard20181** [reply]
This should be removed when the "Dex Optimizer Wrapper" becomes "supported".

**Reply by bengalih** [reply]
My LSposed Home screen reads "Dex Optimizer Wrapper: Supported"
If you mean something else by this, please provide some more specific info.  It isn't helpful to either of us to keep going back and forth like this.

You state that this is somehow based on the determination of SELinux.  And that the determination is based "the actual source code of the implementation."  But the app that is making this determination is Momo, which AFAIK is closed source and therefore I cannot determine why it issues this ART warning only after LSposed is installed.

But, it is clear that LSposed is triggering this warning, not only by its existence, because the error persists when LSposed is removed (and root is also fully removed).  So, once again it is apparent that LSposed is making a change to the system that Momo detects.

I accept that this issue may not be avoidable while LSPosed is running, but the removal of LSPosed should revert the system back to the prior state and, in this case, it is not.

**Reply by Howard20181** [reply]
It has been determined that changes will be restored by the system, please check for changes made by other modules

**Reply by bengalih** [reply]
I have tested this 6 times tonight after doing a full factory reset on the phone.
After a full reset I install Magisk Official and Momo.
I test with Momo and the ART message does not appear.
I install LSposed and reboot and the ART message appears.
I remove LSposed ant the ART message remains.

It is 100% absolutely being caused by the installation of LSposed, and the removal of LSposed does not remove the error.

I am happy to provide any type of logging or other information that would assist, but I have been very methodical in determining the cause.

**Reply by Howard20181** [reply]
Checking other modules
For example, EDXposed, Dreamland, Taichi

**Reply by bengalih** [reply]
None.  As I mentioned I have installed this on a stock Pixel ROM following a clean factory flash.
None of these module exist.  The only thing that gets installed is Magisk (no modules) and Momo.
After LSposed is installed, the ART error appears and will not go away even after removing LSposed .
It does not even go away after reflashing boot.img which removes Magisk.
Only fix is to do another clean factory flash.
Same outcome 5-6 times tonight in testing.

**Reply by Howard20181** [reply]
No fixes can be applied
Please give feedback to AOSP

**Reply by yujincheng08** [reply]
art parameter will be reset after the removal of lsposed and a reboot. if the art paramter persist, nothing can be done from our side.

**Reply by bengalih** [reply]
The installation of LSposed introduces the error, and the removal of it does not remove the error.
This can be recreated at will.

Only the LSposed team can provide information as to what exactly Momo is complaining about with the ART error.

The correct way to resolve this is to disclose what Momo is checking in order to generate this error.

Then, these appropriate settings can be validated before and after LSposed install and removal.

I don't see what AOSP has to do with any of it.

**Reply by bengalih** [reply]
Thank you for your reply. 
So you confirm that an ART parameter is being modified because of the LSposed install?
If the removal of LSPosed does not revert this parameter, can it be done manually?

Can you disclose what parameter is being modified so that it can be manually checked.

I think the general understanding is that if the LSposed module is removed, any changes made by the module itself are also removed.
It sounds like you are saying that is not the case and blaming it on the OS/ROM when this can not be done?

I just would like clarification here because if simply installing LSposed can issue destructive changes that are not reverted on a removal of the module, then the user community should be aware.

**Reply by bengalih** [reply]
So I want to give you one additional piece of information.
I think that it is possible that LSposed is not the guilty party here, but Momo.
My understanding is your team also supports Momo, so I will continue to post here.
If this is not the case, please let me know.

The problem is very strange, and I do not exactly know how to explain it other than to explain the process.
Again, I am doing this on a stock A11 Pixel 3XL ROM after a fresh factory flash.
The only other thing that was installed was Magisk 26.4 (no modules apart from LSposed and Shamiko).

The process is this after installing Momo 4.4.1


1. Momo does not report an ART error
2. After installing Lsposed and Shamiko, Momo still does NOT report the ART error
4. Do an extra reboot to confirm no error
5. Uninstall Momo and reboot
6. Reinstall Momo and **ART error is reported**!!  (NB: that the behavior is the same even without a reboot)
7. Remove Lsposed, Shamiko and reboot
8. Momo still **reports ART error!**
9. Clear Momo cache and reboot again
10. Momo **still reports ART error!**
11. Remove Momo and reinstall
12. Momo does **NOT** report ART error
13. Reinstall LSposed (and Shamiko) and reboot
14. Momo does NOT report ART error

I cannot explain this behavior.  Clearly LSPosed being installed influences Momo's behavior, but following these steps it is possible to create the ART error or remove it based more on when and how Momo is installed.

I cannot continue to do this testing from clean device since I need to finish configuring my phone.

I hope that this information is useful to you since there is clearly a very strange condition with the Momo's app interoperation that I hope you can resolve.

Thank you for your attention to the matter.

**Reply by bengalih** [reply]
After some more tests, in addition to the above, I think the results with Momo can be more succinctly stated with the following:

1. If Momo is installed PRIOR to LSposed being active, it will NOT report the ART error.
2. However, if you install Momo AFTER LSposed, it will display the ART error (this includes a reinstall or upgrade of Momo).*
3. If the error is generated by # 2, it will not go away even after LSposed is removed.
4. The only way to remove the error is to disable LSposed, uninstall Momo, reboot, and install Momo prior to reactivating LSposed.

_* This also includes removing the app and reinstalling with a Swift backup (which technically should be restoring all data)._

So, once again this would show that the issue is with Momo.
While LSposed is the cause of Momo misbehaving, it is Momo which is providing inconsistent results.

Thank you.

**Reply by heximing** [reply]
I have the same issue with @bengalih and the above information is very helpful. At least I understand now the issue is Momo.

However, it is interesting that Momo knows LSPosed existed when we install Momo (Momo knows LSPosed existed to trigger ART error; otherwise, Momo will not trigger ART error, i.e. somehow Momo knows LSPosed existed even if I use `Dr-TSNG/Hide-My-Applist`).

I use Pixel 6 with the official Android 13.0.0 (TQ3A.230901.001, Sep 2023), LSposed 1.9.2 (7024), Magisk 26.4 (26400), Shamiko 0.7.4 (188), Momo 4.4.1.

~One thing I do not understand is, why my LSposed says `Dex Optimizer Wrapper Unsupported (SELinux is permissive)`? How to make it enforced?~ Never mind. rebooting the phone will make `Dex Optimizer Wrapper Supported`.
